src/engine/effects/engineeffectsmanager.cpp

```diff
@@ -130,30 +137,149 @@ void EngineEffectsManager::onCallbackStart() {
+                 numSamples, sampleRate, groupFeatures,
+                 oldGain, newGain);
+}
+
+void EngineEffectsManager::processPostFaderAndMix(
+    const ChannelHandle& inputHandle,
+    const ChannelHandle& outputHandle,
+    CSAMPLE* pIn, CSAMPLE* pOut,
+    const unsigned int numSamples,
+    const unsigned int sampleRate,
+    const GroupFeatureState& groupFeatures,
+    const CSAMPLE_GAIN oldGain,
+    const CSAMPLE_GAIN newGain) {
+    processInner(m_postFaderRacks,
+                 inputHandle, outputHandle,
+                 pIn, pOut,
+                 numSamples, sampleRate, groupFeatures,
+                 oldGain, newGain);
+}
+
+void EngineEffectsManager::processInner(
+    const QList<EngineEffectRack*>& racks,
+    const ChannelHandle& inputHandle,
+    const ChannelHandle& outputHandle,
+    CSAMPLE* pIn, CSAMPLE* pOut,
+    const unsigned int numSamples,
+    const unsigned int sampleRate,
+    const GroupFeatureState& groupFeatures,
+    const CSAMPLE_GAIN oldGain,
+    const CSAMPLE_GAIN newGain) {
+    if (pIn == pOut) {
+        // Gain and effects are applied to the buffer in place,
+        // modifying the original input buffer
+        SampleUtil::applyRampingGain(pIn, oldGain, newGain, numSamples);
+        for (EngineEffectRack* pRack : racks) {
+            if (pRack != nullptr) {
+                pRack->process(inputHandle, outputHandle,
+                               pIn, pIn,
+                               numSamples, sampleRate, groupFeatures);
+            }
+        }
+    } else {
+        // Do not modify the input buffer.
+        // 1. Copy input buffer to a temporary buffer
+        // 2. Apply gain to temporary buffer
+        // 2. Process temporary buffer with each effect rack in series
+        // 3. Mix the temporary buffer into pOut
+        //    ChannelMixer::applyEffectsAndMixChannels(Ramping) use
+        //    this to mix channels into pOut regardless of whether any effects were processsed.
+        int racksProcessed = 0;
+        CSAMPLE* pIntermediateInput = m_buffer1.data();
+        if (oldGain == CSAMPLE_GAIN_ONE && newGain == CSAMPLE_GAIN_ONE) {
+            // Avoid an unnecessary copy. EngineEffectRack::process does not modify the
+            // input buffer when its input & output buffers are different, so this is okay.
+            pIntermediateInput = pIn;
+        } else {
+            SampleUtil::copyWithRampingGain(pIntermediateInput, pIn,
+                                            oldGain, newGain, numSamples);
+        }
+
+        CSAMPLE* pIntermediateOutput = m_buffer2.data();
+        for (EngineEffectRack* pRack : racks) {
+            if (pRack != nullptr) {
+                if (pRack->process(inputHandle, outputHandle,
+                                   pIntermediateInput, pIntermediateOutput,
+                                   numSamples, sampleRate, groupFeatures)) {
+                    ++racksProcessed;
+                    if (racksProcessed % 2) {
```

Comment:
before processing:
// Select an unused intermediate buffer for the next output
CSAMPLE* pIntermediateOutput;
if (pIntermediateInput == m_buffer1.data()) {
   pIntermediateOutput = m_buffer2.data();
} else {
   pIntermediateOutput = m_buffer1.data();
}

after processing
// The output of the previous stage becomes the input of the next stage
pIntermediateInput = pIntermediateOutput;

Reply:
Thank you, this is more readable.